libbeat/registry/backend/memlog/tx.go

```
@@ -0,0 +1,414 @@
+	if checkpoint {
+		tbl = tbl.copySpline()
+		tx.mem.CommitTo(&tbl, true)
+		err := diskStore.commitCheckpoint(&tbl)
+		if err != nil {
+			return err
+		}
+
+		// Acquire exclusive lock -> no readers/writers
+		pending.Lock()
+		defer pending.Unlock()
+		exclusive.Lock()
+		defer exclusive.Unlock()
+
+		// replace hashtable with new snapshot table
+		tx.store.mem.tbl = tbl
+		return nil
+	}
+
+	// Transaction log only mode.
+	// Write transaction state to the transaction log file.
+	// On success we commit all state updates to the in-memory state.
+
+	// block any new readonly transactions being generated
+	pending.Lock()
+	defer pending.Unlock()
+
+	// Append transaction state to the log file + commit to in memory state.
+	err := diskStore.commitOps(&tx.state)
+	if err != nil {
+		return err
+	}
+
+	// wait for all active transaction to be finished before copying the new
+	// state into the shared in memory k/v tables.
+	exclusive.Lock()
+	defer exclusive.Unlock()
+
+	tx.mem.Commit()
+	return nil
+}
+
+func (tx *transaction) Has(k backend.Key) (has bool, err error) {
+	if err = tx.checkRead(); err == nil {
+		has = tx.mem.Has(k)
+	}
+	return
+}
+
+func (tx *transaction) Get(k backend.Key) (v backend.ValueDecoder, err error) {
+	if err = tx.checkRead(); err != nil {
+		return nil, err
+	}
+
+	if tx.IsReadonly() {
+		if vd := tx.mem.GetReadonly(k); vd != nil {
+			return vd, nil
+		}
+		return nil, nil
+	}
+
+	if entry := tx.mem.Get(k); entry != nil {
+		return entry, nil
+	}
+	return nil, nil
+}
+
+func (tx *transaction) Set(k backend.Key, from interface{}) (err error) {
+	if err = tx.checkWrite(); err == nil {
+		err = tx.mem.Set(k, from)
+	}
+	return
+}
+
+func (tx *transaction) Remove(k backend.Key) (err error) {
+	if err = tx.checkWrite(); err == nil {
+		tx.mem.Remove(k)
+	}
+	return
+}
+
+func (tx *transaction) Update(k backend.Key, fields interface{}) (err error) {
+	if err = tx.checkWrite(); err == nil {
+		tx.mem.Update(k, fields)
```

we're ignoring error

---

Indeed.